### PR TITLE
SE(film) support for cudnn backend

### PR DESCRIPTION
**Pull request by ankan-ban**
- split cudnn backend into smaller/more manageable files.
 -- see /src/neural/cuda/readme.txt for more details.
- add support for SE (SE-FiLM)
 -- note that the current SE implementation isn't very fast (causes 20% perf slowdown for 256x20 network on fp16, and even more for smaller networks). 
 -- I am working on optimising it by trying to use a single kernel for the entire SE operation.
     (I had something working for plain SE, which had only 7% perf impact, but that approach won't work for SE-FiLM)

### CODE REVIEW

**Comment by ankan-ban** [comment]
I have created a separate PR for just the refactoring part. https://github.com/LeelaChessZero/lc0/pull/501
It probably makes more sense to merge that first, then merge Ttl's PR https://github.com/LeelaChessZero/lc0/pull/459, and then finally merge this one.

**Comment by borg323** [comment]
Now that #501 and #459 are in, time to move this forward. https://github.com/borg323/lc0/commit/2e109c29e45e1da7311895a2e2bdc304deec3c28 is an effort to merge with master, untested but compiles. Will only be able to test it in about 12 hours.

**Comment by borg323** [comment]
https://github.com/borg323/lc0/commit/8bf7729466378dfc5384fe0de27eb493d9847006 fixes some merge oversights.

**Comment by borg323** [comment]
https://github.com/borg323/lc0/commit/4528caed2c8824b16c68b1d26c986607a6652d39 is an additional commit to fix warnings, I didn't want to merge this on the first round to minimize the diff.

Also I see there are two TODOs still in the code.

**Comment by ankan-ban** [comment]
Thanks @borg323 for merging!

>>Also I see there are two TODOs still in the code.
The TODOs are not strictly blocking deployment of SE-Film. They are for:
1. Optimising SE layer code using single kernel for fp32 mode. (I have a kernel, but it still has some bugs).
2. Finding some way to avoid hardcoding supported channel counts for the optimized SE layer kernel. The kernel is templated on channel counts and right now I have hardcoded instantiations for the channel counts we care about, but it would be good to make it more general.

File: src/neural/cuda/common_kernels.cu [hunk 1]
```diff
@@ -281,6 +281,166 @@ void expandPlanes_Fp16_NHWC(half* output, const uint64_t* masks,
...
+__global__ void globalScale_kernel(float* output, const float* input,
+                                   const float* scaleBias, int inputSize,
+                                   int C) {
+  const int kPlaneSize = 64;
+
+  int tid = blockIdx.x * blockDim.x + threadIdx.x;
+
+  if (tid > inputSize) return;
+
+  float val1 = input[tid];   // skip connection
```
[the remark this answers is not in the document]
Comment (ankan-ban) [reply src/neural/cuda/common_kernels.cu]: Ok, I have changed most of the comments in all files of cuda backend.
Will do another pass for to fix variable naming in all places (lot of them). Maybe separate PR ?

File: src/neural/cuda/layers.cc [hunk 2]
```diff
@@ -263,6 +263,128 @@ BNLayer<DataType>::~BNLayer() {
...
+void SELayer<half>::LoadWeights(float* w1, float* b1, float* w2, float* b2,
+                                float* prevLayerBias, void* scratch) {
+  size_t num_weights1 = C * numFc1Out_;
+  size_t weight_size1 = sizeof(float) * num_weights1;
+
+  size_t num_weights2 = 2 * num_weights1;
+  size_t weight_size2 = 2 * weight_size1;
+
+  // transpose the weight matrices
+  float* temp = new float[weight_size2];
```
[the remark this answers is not in the document]
Comment (ankan-ban) [reply src/neural/cuda/layers.cc]: Fixed, actually no reason at all. 
Not accustomed of using std lib much :-/

**Comment by ankan-ban** [comment]
I did some simplification/refactoring of SE related code, and fused the bias addition for fp32 inside kernels of SE layer. Right now fp16 has a single fused kernel, but I wasn't able to get any improvement for fp32 with similar approach (mostly because my matrix multiply part of the kernel is too slow compared to cublas). So, fp32 is still using 4 passes. 

Both fp16 and fp32 are about 10% slower than non-SE network of same size. Here are the performance nos for a 10x128 network:
`Raw NPS for batch size of 256 on TU102

         Baseline          SE-Film
FP16:    107098            97235.7
FP32:    39695.6           36318.1 `
20x256 network is expected to have smaller impact in performance.

**Pull request merged**